feat: btc-staker-improvement-107

wnjoon opened the pull request:
This PR indicates the implementation of issue #107

For a quick review, we create a PR that includes the completed tasks: `delete watch-staking` and `remove post-approval`.

## Delete `watch-staking` endpoint

- Removed related RPC call paths and functions
- Eliminated the `watchTxData` field and related logic in `stakingRequestCmd`
- Removed database fields and query/update logic related to watched transactions
- Modified message fields accordingly
- Cleaned up functions that previously had a single execution branch after deleting watched (e.g., `buildOwnDelegation`)
- Removed test codes related to watch-staking

As a result, calling watch_staking_tx RPC will output the following:

```sh
# Run stakerd, babylon, and bitcoind binaries by referring to the contents of README.md. 
# The following shows the log of stakerd.
$ ./stakerd
time="2025-01-31T20:06:59+09:00" level=info msg="Starting App"
time="2025-01-31T20:06:59+09:00" level=info msg="Connecting to node backend: bitcoind"
time="2025-01-31T20:06:59+09:00" level=info msg="Successfully started Prometheus metrics server at 127.0.0.1:2112"
time="2025-01-31T20:06:59+09:00" level=info msg="Successfully connected to node backend: bitcoind"
time="2025-01-31T20:06:59+09:00" level=info msg="Initial btc best block height is: 233384"
time="2025-01-31T20:06:59+09:00" level=info msg="App started"
time="2025-01-31T20:06:59+09:00" level=info msg="Staker Service fully started"
time="2025-01-31T20:06:59+09:00" level=info msg="I[2025-01-31|20:06:59.435] serve                                        msg=\"Starting RPC HTTP server on 127.0.0.1:15812\""

# Call watch_staking_tx RPC
$ curl -X POST -H "Content-Type: application/json" http://127.0.0.1:15812 -d '{"jsonrpc":"2.0","id":1,"method":"watch_staking_tx","params":["","1000","100000000","",[],"","","","","","","","","1000","1"]}'

# Message "Method not found" is returned
{"jsonrpc":"2.0","id":1,"error":{"code":-32601,"message":"Method not found"}}
```

## Remove `post-approval` flow

- Removed the `sendToBabylonFirst` option while staking and its related codes
- Removed the `delegationActivatedPostApprovalEvent` and its related codes and channel
- Removed the `handlePostApprovalCmd` function and simplifying call hierarchy
- Removed `delegationActive` field in `unbondingTxSignaturesConfirmedOnBabylonEvent`
- Removed `usePreApprovalFlow` field in `stakingRequestCmd`
- Removed post-approval-related channels and events, and simplified the `checkTransactionsStatus` function.
- Modified e2e tests that used post-approval logic to use pre-approval logic.

KonradStaniec commented on itest/e2e_test.go:
```diff
@@ -39,122 +39,6 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestStakingFailures(t *testing.T) {
```
I would still leave this test as it test few external failues in e2e setting.

wnjoon replied:
fixed [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

KonradStaniec commented on itest/e2e_test.go:
```diff
@@ -263,74 +147,6 @@ func TestSendingStakingTransactionWithPreApproval(t *testing.T) {
-		testStakingData3,
-		testStakingData4,
-		testStakingData5,
-	}, false)
```
I would leave this test, but use pre-approval flow to create initial stake

wnjoon replied:
fixed [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

KonradStaniec commented on itest/e2e_test.go:
```diff
@@ -386,242 +202,6 @@ func TestMultiplePreApprovalTransactions(t *testing.T) {
-	}
-}
-
-func TestStakingUnbonding(t *testing.T) {
```
this test is also needed i.e we need to first stake with pre-approval flow, activate delegation and then unbond

wnjoon replied:
fixed [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

KonradStaniec commented on itest/e2e_test.go:
```diff
@@ -739,79 +319,6 @@ func TestBitcoindWalletBip322Signing(t *testing.T) {
 	require.NoError(t, err)
 }
 
-func TestRecoverAfterRestartDuringWithdrawal(t *testing.T) {
```
this test is also needed but should be done with pre-approval flow instead

wnjoon replied:
fixed [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

KonradStaniec reviewed (commented):
In general looks good. Some general remarks:
- If I remember correctly, after dropping support for post approval flow we can remove `app.stakingTxBtcConfirmedEvChan` and `app.delegationSubmittedToBabylonEvChan` event channels (and corresponding events)
- we should be also able to simplify function which handles recovery after restart i.e `func (app *App) checkTransactionsStatus()` as now some states are not need

wnjoon commented:
@KonradStaniec 

I have a question regarding the removal of the post-approval logic and the changes to the database structure.

First, the pre-approval process I reviewed is as follows:

1. The staking transaction is sent to Babylon, and its initial status is PENDING.
2. Once quorum or more covenant signatures are completed, the status changes to VERIFIED.
3. If it is confirmed that the transaction sent to Babylon has been included in a Bitcoin block, the status changes to ACTIVE.

Ultimately, `btc-staker` will store all staking transactions in the database, but the status can be directly verified in both Babylon and Bitcoin.

There seem to be two cases where a resend in `checkTransactionsStatus` may be necessary:
- When the transaction is VERIFIED in Babylon but does not exist in Bitcoin.
- When the transaction is ACTIVE in Babylon but does not exist in Bitcoin.

Since each status appears to be directly verifiable on-chain, I would like to first confirm whether this structure is correct.

KonradStaniec commented:
> First, the pre-approval process I reviewed is as follows:
> 
> The staking transaction is sent to Babylon, and its initial status is PENDING.
> Once quorum or more covenant signatures are completed, the status changes to VERIFIED.
> If it is confirmed that the transaction sent to Babylon has been included in a Bitcoin block, the status changes to ACTIVE.

Yes, this is correct understanding of the pre-approval flow. There are some details in there which I can fill:
- only after status is changed to Verified, staker send transaction to BTC 
- after sending the BTC we wait for other process to report the inclusion to Babylon, and only after transaction is Active on Babylon we change internal status to Active

> Ultimately, btc-staker will store all staking transactions in the database, but the status can be directly verified in both Babylon and Bitcoin.

yes :) 

> There seem to be two cases where a resend in checkTransactionsStatus may be necessary:
> 
> When the transaction is VERIFIED in Babylon but does not exist in Bitcoin (TxInMemPool).
> 

Yep, it we restart the program at any point between sending to Babylon and sending staking tx to BTC, we need to somehow recover:
- after restart if transactions is pending, we need to start checking Babylon until it gets verified and then procceed with rest of the flow
- after restart if transactions is verified, we need to check if its on BTC and if not send it there.

There are also recovery conditions related to unbonding i.e
- after restart if state is active, we need to check if staking output is still unspent and tx is still active on Babyln

wnjoon commented:
> > First, the pre-approval process I reviewed is as follows:
> > The staking transaction is sent to Babylon, and its initial status is PENDING.
> > Once quorum or more covenant signatures are completed, the status changes to VERIFIED.
> > If it is confirmed that the transaction sent to Babylon has been included in a Bitcoin block, the status changes to ACTIVE.
> 
> Yes, this is correct understanding of the pre-approval flow. There are some details in there which I can fill:
> 
> * only after status is changed to Verified, staker send transaction to BTC
> * after sending the BTC we wait for other process to report the inclusion to Babylon, and only after transaction is Active on Babylon we change internal status to Active
> 
> > Ultimately, btc-staker will store all staking transactions in the database, but the status can be directly verified in both Babylon and Bitcoin.
> 
> yes :)
> 
> > There seem to be two cases where a resend in checkTransactionsStatus may be necessary:
> > When the transaction is VERIFIED in Babylon but does not exist in Bitcoin (TxInMemPool).
> 
> Yep, it we restart the program at any point between sending to Babylon and sending staking tx to BTC, we need to somehow recover:
> 
> * after restart if transactions is pending, we need to start checking Babylon until it gets verified and then procceed with rest of the flow
> * after restart if transactions is verified, we need to check if its on BTC and if not send it there.
> 
> There are also recovery conditions related to unbonding i.e
> 
> * after restart if state is active, we need to check if staking output is still unspent and tx is still active on Babyln

How should we handle the `ACTIVATED` status? I’m unsure whether we need to check Bitcoin to verify if the staking transaction is in the mempool or confirmed.

In the `activateVerifiedDelegation` function, even when the queried status from Babylon is `ACTIVE`(from db in btc-staker), it still checks whether the transaction exists in Bitcoin.

This is confusing because I assumed that the `ACTIVE` status from btc-staker(and `ACTIVATE` status from Babylon) guarantees that the staking transaction is already in Bitcoin and that the required block confirmations have been met.

Based on your comment, the modified logic could be structured as follows:

1. First, check the status from Babylon.
2. If the transaction doesn’t exist, send the delegation transaction to Babylon.
3. If the status is `PENDING`, keep checking Babylon until it becomes `VERIFIED`, then proceed with the remaining steps (I believe the `checkForUnbondingTxSignaturesOnBabylon` function is related).
4. If the status is `VERIFIED`, check Bitcoin for the transaction. If it doesn’t exist, send it to Bitcoin.

And my opinion is:

5. If the status is `ACTIVATED`, no further action is needed. 
6. There’s no need to check Bitcoin independently without first checking Babylon, as the staking transaction must be sent to Babylon before being broadcast to Bitcoin.

wnjoon commented:
Comment Summary [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

- Removed `post-approval` related events and channels (ref: [link](https://github.com/babylonlabs-io/btc-staker/pull/128#pullrequestreview-2591923398))
- Simplified `checkTransactionsStatus` function by removing `post-approval` related functions
- Fixed e2e-test for using `pre-approval` logic

wnjoon commented:
> In general looks good. Some general remarks:
> 
> * If I remember correctly, after dropping support for post approval flow we can remove `app.stakingTxBtcConfirmedEvChan` and `app.delegationSubmittedToBabylonEvChan` event channels (and corresponding events)
> * we should be also able to simplify function which handles recovery after restart i.e `func (app *App) checkTransactionsStatus()` as now some states are not need

Fixed [151b234](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/151b234e61f1b030cf0bb87f084da46c2bee12c7)

KonradStaniec commented on staker/stakerapp.go:
```diff
@@ -575,40 +443,30 @@ func (app *App) SendPhase1Transaction(
 		switch tx.State {
 		case proto.TransactionState_SENT_TO_BTC:
-			transactionsSentToBtc = append(transactionsSentToBtc, &stakingTxHash)
+			// TODO: nothing to do, should be removed in proto
```
do you think it is better to remove this states in this pr or in subsequent one about reducing the state in the program ?

wnjoon replied:
@KonradStaniec 
I currently believe that if we retrieve the status of transactions directly from Bitcoin, there will be no need to store Bitcoin’s status. Can I give it a try and update you in the next commit?

KonradStaniec commented on staker/events.go:
```diff
@@ -12,51 +11,14 @@ type StakingEvent interface {
-var _ StakingEvent = (*stakingTxBtcConfirmedEvent)(nil)
-var _ StakingEvent = (*delegationSubmittedToBabylonEvent)(nil)
-var _ StakingEvent = (*delegationActivatedPostApprovalEvent)(nil)
 var _ StakingEvent = (*delegationActivatedPreApprovalEvent)(nil)
```
now that we do not have `postApproval` flow , maybe we can rename it `delegationActivatedEvent` ?

wnjoon replied:
fixed [2da7401](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/2da7401dbd5fdb4ee828902c194334d388c73e8b)

wnjoon commented:
Comment Summary [2da7401](https://github.com/babylonlabs-io/btc-staker/pull/128/commits/2da7401dbd5fdb4ee828902c194334d388c73e8b)

@KonradStaniec 
I tried to split the work and commit it, but since many parts are interconnected, many changes were made. I appreciate your understanding. 🙏

- Changed the process to query the status of transactions sent to Babylon directly from the chain instead of the local database. 
- Since the implementation for checking the status directly from Bitcoin has not been completed yet, the Bitcoin status is still checked locally. To address this, we modified the existing `staking_details` RPC to retrieve the status directly from the Babylon node and added `stored_staking_details` to check the status locally.
- Modified the name and logic of the `checkTransactionsStatus` function. The reason for renaming it is that the function performs follow-up processing based on the checked status. The status of Babylon transactions is retrieved directly from Babylon, while Bitcoin transactions are still checked locally.
- All transactions are stored in the local database via the `AddNewStoredTransaction` function after being successfully sent to Babylon. Since the status is no longer stored in the local database, test cases related to status transitions have been modified or removed.
- Modified some functions to wrap errors using the fmt.Errorf syntax, but not all functions were changed.

wnjoon commented:
@KonradStaniec 

While working on the current PR, I realized that the workload was too large, and the PR was becoming much bigger than expected.

To manage this, I plan to split the changes into two phases:
- Phase 1 PR: Includes all modifications except for the parts that update the status in the database.
- Phase 2 PR: Covers the remaining changes, which I am already working on and will submit as a separate PR shortly.
